internal/planner/compatible.go

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package planner
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// IncompatibleInstanceError indicates SmbShare resources are not compatible
+// with each other and can not hosted by the same server.
+type IncompatibleInstanceError struct {
+	current  string
+	existing string
+	reason   string
+}
+
+// Error interface method.
+func (iie IncompatibleInstanceError) Error() string {
+	return fmt.Sprintf("Share resource %s is incompatible with %s: %s",
+		iie.current,
+		iie.existing,
+		iie.reason)
+}
+
+func incompatible(
+	current, existing InstanceConfiguration,
```

Comment:
I prefer using pass-by-pointer in such cases. Although `InstanceConfiguration` is just a 4-pointers tuple now, it may get bigger in the future causing inefficient pass-by-value.

Reply:
This is a case where I personally feel semantics > performance UNTIL we profile and discover this is leading to a measurable performance issue. I want the "read only" nature of this comparison to be apparent.

On top of that https://codewithyury.com/golang-pass-by-pointer-vs-pass-by-value/ notes that pass by value is not always slower. I one read a longer and more thorough article saying the same but I can't find it right now.